lib/Fencer/Types.hs

```diff
@@ -193,10 +215,9 @@ type RuleTree = HashMap (RuleKey, Maybe RuleValue) RuleBranch
 
 -- | A single branch in a rule tree, containing several (or perhaps zero)
 -- nested rules.
-data RuleBranch = RuleBranch
-    { ruleBranchRateLimit :: !(Maybe RateLimit)
-    , ruleBranchNested :: !RuleTree
-    }
+data RuleBranch
+  = RuleBranch {  ruleBranchNested :: !RuleTree }
+  | RuleLeaf { ruleBranchRateLimit :: !RateLimit }
```

Comment:
Again, I'd remove partial fields and turn them into haddocks. Whenever you want to extract a value from a sum type, you have to handle the other alternatives and partial fields sweep that requirement under the carpet, which can easily lead to bugs.

Reply:
The same argument from above applies here. I must be missing something because otherwise it looks to me like you're saying that sum types are to be avoided.